### PR TITLE
MLv2 implement `suggested-join-condition`

### LINKED ISSUE #31175: [MLv2] Implement/expose `suggestedJoinCondition` function
When building a join against a table with a known FK->PK relationship (the source table has an FK that points to the PK of the table to be joined), in current MLv1 behavior the condition is automatically populated with those columns. Examples:

###### Reviews -> Products

image

Note that this only seems to work with the Source Table (or previous stage) has an FK to the PK of the table. No such default condition is applied when flipping the tables around, for example:

###### Products -> Reviews

image

Or if a previously-joined Table has the FK, rather than the source Table:

###### Products -> Orders -> People

image

### Signature

I think the appropriate signature for this function should be something like

```ts
export function suggestedJoinCondition(query: Query, stageNumber: integer, tableToJoin: TableMetadata): Clause? { ... }
```

and should return an MBQL clause representing the suggested join condition if one exists, otherwise `nil`/`null`/etc. if no such suggested condition exists.

Note that while we could make this a little smarter and handle PK -> FK as well I think for now we should just try to replicate the old behavior.

### PR DESCRIPTION
**Pull request by camsaul**
Resolves #31175

### CODE REVIEW

File: src/metabase/lib/join.cljc [hunk 1]
```diff
@@ -455,9 +470,39 @@
...
+                (lib.metadata.calculation/visible-columns query stage-number x)))
+
+(mu/defn ^:private fk-column :- [:maybe lib.metadata/ColumnMetadata]
+  "Given a query stage find an FK column that points to the PK `pk-col`."
+  [query        :- ::lib.schema/query
+   stage-number :- :int
+   pk-col       :- [:maybe lib.metadata/ColumnMetadata]]
+  (when-let [pk-id (:id pk-col)]
+    (m/find-first (fn [{:keys [semantic-type fk-target-field-id], :as _col}]
+                    (and (isa? semantic-type :type/FK)
```
Comment (bshepherdson) [inline src/metabase/lib/join.cljc:491]: I wonder if we should prefer the functions like `(metabase.lib.types.isa/foreign-key? col)` over direct checks like this. They are overridable should the need arise.

File: test/metabase/lib/join_test.cljc [hunk 2]
```diff
@@ -418,3 +418,42 @@
...
+                      (meta/table-metadata :categories)))
+      ;; plain query
+      (lib/query meta/metadata-provider (meta/table-metadata :venues))
+
+      ;; query with an aggregation (FK column is not exported, but is still "visible")
+      (-> (lib/query meta/metadata-provider (meta/table-metadata :venues))
+          (lib/aggregate (lib/count))))))
+
+(deftest ^:parallel suggested-join-condition-pk->fk-test
+  (testing "Don't suggest join conditions for a PK -> FK relationship"
```
Comment (metamben) [inline test/metabase/lib/join_test.cljc:439]: What's the reasoning behind this? Isn't this some arbitrary asymmetry?

File: src/metabase/lib/join.cljc [hunk 3]
```diff
@@ -455,9 +470,39 @@
...
-    {:lib/type :mbql.filter/operator, :short :!=, :display-name (i18n/tru "Not equal to")}]))
+   (join-condition-operator-definitions)))
+
+(mu/defn ^:private pk-column :- [:maybe lib.metadata/ColumnMetadata]
+  "Given something `x` (e.g. a Table metadata) find the PK column."
+  [query        :- ::lib.schema/query
+   stage-number :- :int
+   x]
+  (m/find-first (fn [{:keys [semantic-type], :as _col}]
+                  (isa? semantic-type :type/PK))
```
Comment (metamben) [inline src/metabase/lib/join.cljc:481]: Shouldn't we use `metabase.lib.types.isa/primary-key`?

File: src/metabase/lib/join.cljc (hunk 1 above)
Comment (metamben) [inline src/metabase/lib/join.cljc:491]: `metabase.lib.types.isa/foreign-key?`

**Review by bshepherdson — APPROVED** [review]

File: src/metabase/lib/join.cljc [hunk 4]
```diff
@@ -455,9 +470,39 @@
...
+  primary key of the thing we're joining (see #31175 for more info); otherwise this will return `nil` if no default
+  condition is suggested."
+  ([query joined-thing]
+   (suggested-join-condition query -1 joined-thing))
+
+  ([query         :- ::lib.schema/query
+    stage-number  :- :int
+    joined-thing]
+   (when-let [pk-col (pk-column query stage-number joined-thing)]
+     (when-let [fk-col (fk-column query stage-number pk-col)]
```
[the remark this answers is not in the document]
Comment (camsaul) [reply src/metabase/lib/join.cljc:507]: It's always seemed kinda weird to me that that's not how `when-let` works by default, when Emacs Lisp copied it from Clojure in 25 they supported multiple bindings e.g.

https://github.com/emacs-mirror/emacs/blob/4b3de748b0b04407d2492500c77905de56de1180/lisp/emacs-lisp/comp-cstr.el#L483-L484 

I'm not against the idea of adding it

File: test/metabase/lib/join_test.cljc (hunk 2 above)
Comment (camsaul) [reply test/metabase/lib/join_test.cljc:439]: idk, it's just how it works in MLv1. I was trying to replicate existing behavior so we didn't break a bunch of tests

File: src/metabase/lib/join.cljc [hunk 5]
```diff
@@ -455,9 +470,39 @@
...
+(mu/defn ^:private pk-column :- [:maybe lib.metadata/ColumnMetadata]
+  "Given something `x` (e.g. a Table metadata) find the PK column."
+  [query        :- ::lib.schema/query
+   stage-number :- :int
+   x]
+  (m/find-first (fn [{:keys [semantic-type], :as _col}]
+                  (isa? semantic-type :type/PK))
+                (lib.metadata.calculation/visible-columns query stage-number x)))
+
+(mu/defn ^:private fk-column :- [:maybe lib.metadata/ColumnMetadata]
```
[the remark this answers is not in the document]
Comment (camsaul) [reply src/metabase/lib/join.cljc:483]: 👍, I'll fix this is in a follow on PR I'm currently working on.

**Pull request merged**